File: admin-app/lib/buying-cost-estimator.ts
export type BuyingCostPurchaseContext = 'thai_local' | 'foreign';
export type BuyingCostOwnershipType = 'freehold' | 'leasehold' | 'company_hold';
export type BuyingCostTransferSplit = 'buyer_pays' | 'split_equally' | 'seller_pays';
export type BuyingCostFinancingMode = 'cash' | 'financing';

export type BuyingCostEstimateRequest = {
  purchase_context: BuyingCostPurchaseContext;
  property_price: number;
  ownership_type: BuyingCostOwnershipType;
  transfer_split: BuyingCostTransferSplit;
  financing_mode: BuyingCostFinancingMode;
  assumption_set_id: string;
  assumption_set_version: string;
  agent_fee?: number;
  lawyer_fee?: number;
  bank_transfer_cost?: number;
  fx_estimate?: number;
};

export type BuyingCostEstimateLineItem = {
  key: string;
  label_key: string;
  amount: number;
  source_type: 'fixed' | 'editable';
  included_in_total: boolean;
};

export type BuyingCostEstimateResponse = {
  assumption_set_id: string;
  assumption_set_version: string;
  purchase_context: BuyingCostPurchaseContext;
  line_items: BuyingCostEstimateLineItem[];
  government_fees: number;
  closing_cost: number;
  total_cash_needed: number;
  unresolved_items: string[];
  disclaimer_key: string;
};

export const DEFAULT_BUYING_COST_ASSUMPTION_SET_ID = 'amp_v2_buying_cost_baseline';
export const DEFAULT_BUYING_COST_ASSUMPTION_SET_VERSION = '2026-03-15';

export async function requestBuyingCostEstimate(
  input: BuyingCostEstimateRequest,
): Promise<BuyingCostEstimateResponse> {
  const response = await fetch('/api/tools/buying-cost', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(input),
  });

  if (!response.ok) {
    throw new Error(`request_failed:${response.status}`);
  }

  return (await response.json()) as BuyingCostEstimateResponse;
}

File: packages/core/schemas/buying_cost_estimator.py
from __future__ import annotations

from decimal import Decimal
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field


class PurchaseContext(str, Enum):
    THAI_LOCAL = "thai_local"
    FOREIGN = "foreign"


class OwnershipType(str, Enum):
    FREEHOLD = "freehold"
    LEASEHOLD = "leasehold"
    COMPANY_HOLD = "company_hold"


class TransferSplit(str, Enum):
    BUYER_PAYS = "buyer_pays"
    SPLIT_EQUALLY = "split_equally"
    SELLER_PAYS = "seller_pays"


class FinancingMode(str, Enum):
    CASH = "cash"
    FINANCING = "financing"


class BuyingCostEstimatorRequest(BaseModel):
    purchase_context: PurchaseContext
    property_price: Decimal = Field(gt=0)
    ownership_type: OwnershipType
    transfer_split: TransferSplit
    financing_mode: FinancingMode
    assumption_set_id: str = Field(min_length=1)
    assumption_set_version: str = Field(min_length=1)
    agent_fee: Decimal | None = Field(default=None, ge=0)
    lawyer_fee: Decimal | None = Field(default=None, ge=0)
    bank_transfer_cost: Decimal | None = Field(default=None, ge=0)
    fx_estimate: Decimal | None = Field(default=None, ge=0)

    model_config = ConfigDict(extra="ignore")


class BuyingCostLineItem(BaseModel):
    key: str
    label_key: str
    amount: Decimal
    source_type: str
    included_in_total: bool = True


class BuyingCostEstimatorResponse(BaseModel):
    assumption_set_id: str
    assumption_set_version: str
    purchase_context: PurchaseContext
    line_items: list[BuyingCostLineItem]
    government_fees: Decimal
    closing_cost: Decimal
    total_cash_needed: Decimal
    unresolved_items: list[str]
    disclaimer_key: str